### PR TITLE
TypeScript rewrite + updates

### PR DESCRIPTION
**Pull request by danrouse**
- Rewritten in TS
- Removed manually-written field definitions
- Added parser for C headers with field definitions for all structs
- Updated work area parsing to use new field definitions

Since the field definitions are all generated from source, all functions are fully supported. However, since this uses field names from those definitions, some keys have changed since the previous version.

This is published to npm as version 2.0.6 for use by the geocoder.

### CODE REVIEW

File: typings/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+type CDeclaration = {
+  name: string;
+  type: 'char' | 'union' | 'struct' | string;
```
Comment (jordanhamill) [inline typings/index.d.ts]: should `string` be quoted here?

File: src/extract-typedefs.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    ...decl,
+    byteLength: getSize(decl, typedefs) * reduceCount(decl.count)
+  };
+  if (decl.children) {
+    next.children = decl.children.map(c => applySize(c, typedefs));
+  }
+  return next;
+}
+
+function parseHeaderFile(path: string) {
```
Comment (jordanhamill) [inline src/extract-typedefs.ts]: Is this home grown? Just wondering if anyone else has solved the problem of generating TS definitions for FFI

Comment (danrouse) [reply src/extract-typedefs.ts]: This is indeed home-grown. I went down quite a rabbit hole with this, trying to pull in various parts of C toolchains in order to parse the header file. As insane as this script is, it seemed like the most reasonable approach after quite a bit of research. Existing tools for analyzing C source code work at low enough of a level that they don't offer a much better experience than just string parsing as is done here.

It's potentially brittle but this is a pretty specific case, so the simplicity of doing string parsing versus traversing an AST seemed worth it.

File: typings/index.d.ts (hunk 1 above)
Comment (danrouse) [reply typings/index.d.ts]: No, the strings in this case are the other type names defined in the header (e.g. `C_WA2_F1A`)

File: src/extract-typedefs.ts (hunk 2 above)
Comment (jordanhamill) [reply src/extract-typedefs.ts]: Yeah it's some scary parsing going on, I guess if it works, it works! It's only generating TS definitions so it's not _too_ critical if it broke down the line.

**Comment by vickeryj** [comment]
is this ready to merge?

**Pull request merged**